src/omp/backend.f90

```diff
@@ -393,23 +362,21 @@ real(dp) function scalar_product_omp(self, x, y) result(s)
 
   end function scalar_product_omp
 
-  subroutine copy_into_buffers(u_send_s, u_send_e, u, n, n_blocks)
+  subroutine copy_into_buffers(u_send_s, u_send_e, u)
     implicit none
 
     real(dp), dimension(:, :, :), intent(out) :: u_send_s, u_send_e
-    real(dp), dimension(:, :, :), intent(in) :: u
-    integer, intent(in) :: n
-    integer, intent(in) :: n_blocks
+    class(field_t), intent(in) :: u
     integer :: i, j, k
     integer :: n_halo = 4
 
     !$omp parallel do
-    do k = 1, n_blocks
+    do k = 1, u%n_groups
       do j = 1, n_halo
         !$omp simd
-        do i = 1, SZ
-          u_send_s(i, j, k) = u(i, j, k)
-          u_send_e(i, j, k) = u(i, n - n_halo + j, k)
+        do i = 1, u%sz
+          u_send_s(i, j, k) = u%data(i, j, k)
+          u_send_e(i, j, k) = u%data(i, u%n - n_halo + j, k)
```

Comment:
In particular here, we can't have `field_t%n`. We have to pass an `n` value into this subroutine which is determined from the corresponding `tdsops`, when `copy_into_buffers` is called in `tds_solve`. `tds_solve` always have a `tdsops`, and passing the `tdsops%n` down to `copy_into_buffers` is necessary. Otherwise we'll end up sending wrong set of slices.